src/Compiler/Service/ServiceAnalysis.fs

```diff
@@ -451,3 +452,1056 @@ module UnusedDeclarations =
+                // Pattern (A :: _)
+                // Pattern (A | B)
+                // Pattern (A & B)
+                // Pattern (A as B)
+                // Pattern (A, B)
+                // Pattern1 (Pattern2 (x = A))
+                // Pattern1 (Pattern2 x y)
+                | SynPat.LongIdent _, SynPat.Typed _
+                | SynPat.LongIdent _, SynPat.Attrib _
+                | SynPat.LongIdent _, SynPat.IsInst _
+                | SynPat.LongIdent _, SynPat.ListCons _
+                | SynPat.LongIdent _, SynPat.Or _
+                | SynPat.LongIdent _, SynPat.Ands _
+                | SynPat.LongIdent _, SynPat.As _
+                | SynPat.LongIdent _, SynPat.Tuple(isStruct = false)
+                | SynPat.LongIdent _, SynPat.LongIdent(argPats = SynArgPats.NamePatPairs _)
+                | SynPat.LongIdent _, SynPat.LongIdent(argPats = SynArgPats.Pats (_ :: _))
+
+                // A | (B as C)
+                // A & (B as C)
+                // A, (B as C)
+                | SynPat.Or _, SynPat.As _
+                | SynPat.Ands _, SynPat.As _
+                | SynPat.Tuple _, SynPat.As _
+
+                // x, (y, z)
+                | SynPat.Tuple _, SynPat.Tuple(isStruct = false)
+
+                // A, (B | C)
+                // A & (B | C)
+                | SynPat.Tuple _, SynPat.Or _
+                | SynPat.Ands _, SynPat.Or _
+
+                // (x : int) | x
+                // (x : int) & y
+                | SynPat.Or _, SynPat.Typed _
+                | SynPat.Ands _, SynPat.Typed _
+
+                // let () = …
+                // member _.M() = …
+                | SynPat.Paren _, SynPat.Const (SynConst.Unit, _)
+                | SynPat.LongIdent _, SynPat.Const (SynConst.Unit, _) -> ValueNone
+
+                | _, SynPat.Const _
+                | _, SynPat.Wild _
+                | _, SynPat.Named _
+                | _, SynPat.Typed _
+                | _, SynPat.LongIdent(argPats = SynArgPats.Pats [])
+                | _, SynPat.Tuple(isStruct = true)
+                | _, SynPat.Paren _
+                | _, SynPat.ArrayOrList _
+                | _, SynPat.Record _
+                | _, SynPat.Null _
+                | _, SynPat.OptionalVal _
+                | _, SynPat.IsInst _
+                | _, SynPat.QuoteExpr _
+
+                | SynPat.Or _, _
+                | SynPat.ListCons _, _
+                | SynPat.Ands _, _
+                | SynPat.As _, _
+                | SynPat.LongIdent _, _
+                | SynPat.Tuple _, _
+                | SynPat.Paren _, _
+                | SynPat.ArrayOrList _, _
+                | SynPat.Record _, _ -> ValueSome range
+
+                | _ -> ValueNone
+
+            | _ -> ValueNone
+
+    let getUnnecessaryParentheses (getSourceLineStr: int -> string) (parsedInput: ParsedInput) : Async<range seq> =
+        async {
+            let ranges = HashSet Range.comparer
+
+            let visitor =
```

Comment:
@psfinaki Hmm, if you're referring specifically to the use of the visitor, then `FSharpParseFileResults` seems like it would be a better fit than `ServiceParseTreeWalk`, although I'd still lean towards how it is here. I was basically following the model of `UnusedOpens.getUnusedOpens`, `SimplifyNames.getSimplifiableNames`, and `UnusedDeclarations.getUnusedDeclarations` in this file, except that those use the check results instead of the parse results.

These modules and their single public functions in `FSharp.Compiler.EditorServices` do seem like they could perhaps be reorganized under modules dedicated to the AST (`FSharpParseFileResults`) and TAST (`FSharpCheckFileResults`), and then either applied to `FSharpParseFileResults` or `FSharpCheckFileResults` values or else exposed as instance methods on those types.

Reply:
Yeah it is part of a bigger problem in a way, we have similar pieces of code all over the place without much guidelines. So if you have an idea how to organize this properly, your fresh view would be helpful.